onnxruntime/contrib_ops/cpu/tokenizer.cc

```diff
@@ -0,0 +1,472 @@
+      new (output_data + output_index) std::string(pad_value_);
+      ++output_index;
+    }
+    if (mark_) {
+      new (output_data + output_index) std::string(&end_text, 1);
+      ++output_index;
+    }
+    ++curr_input;
+  }
+  return Status::OK();
+}
+
+Status Tokenizer::SeparatorTokenize(OpKernelContext* ctx,
+                                    size_t N, size_t C,
+                                    const std::vector<int64_t>& input_dims) const {
+  using namespace tokenizer_details;
+
+  std::wstring_convert<std::codecvt_utf8<wchar_t>> converter(conv_error, wconv_error);
+  // Scan all strings and attempt to find separators in them
+  // collect all the output tokens here
+  size_t max_tokens = 0;
+  std::vector<std::vector<std::wstring>> tokenized_strings;
+  tokenized_strings.reserve(N * C);
+  auto X = ctx->Input<Tensor>(0);
+  auto const input_data = X->template Data<std::string>();
+  auto curr_input = input_data;
+  auto const last = input_data + N * C;
+  while (curr_input != last) {
+    const auto& s = *curr_input;
+    std::wstring wstr = converter.from_bytes(s);
+    if (wstr == wconv_error) {
+      return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT,
+                    "Invalid utf8 chars in the input: " + s);
+    }
+
+    struct Match {
+      int priority_;
+      size_t offset_;
+      size_t size_;
+      // create a conflict for overlapping matches
+      // thus if they overlap neither is less than the other
+      // and they are considered equal
+      bool operator<(const Match& o) const {
+        return (offset_ + size_) <= o.offset_;
+      }
+    };
+
+    std::set<Match> matches;
+    const wchar_t* ws = wstr.c_str();
+    size_t len_remaining = wstr.length();
+    size_t offset = 0;
+    while (len_remaining > 0) {
+      const auto* val = search_data_->tst_.get(ws, len_remaining);
+      if (val != nullptr) {
+        auto p = matches.insert({val->priority_, offset, val->w_len});
+        while (!p.second && val->priority_ < p.first->priority_) {
+          // if overlapping matches of the same pattern(priority), then
+          // the earlier match naturally wins
+          matches.erase(p.first);
+          p = matches.insert({val->priority_, offset, val->w_len});
+        }
+      }
+      ++ws;
+      ++offset;
+      --len_remaining;
+    }
+
+    // Tokenize
+    tokenized_strings.emplace_back();
+    auto& row_tokens = tokenized_strings.back();
+    row_tokens.reserve(matches.size() + 1);
+    ws = wstr.c_str();
+    offset = 0;
+    for (const auto& m : matches) {
+      assert(m.offset_ >= offset);
+      size_t sz = (m.offset_ - offset);
+      if (sz > 0 && sz >= size_t(mincharnum_)) {
+        row_tokens.emplace_back(ws, sz);
+      }
+      offset = m.offset_ + m.size_;
+      ws = wstr.c_str() + offset;
+    }
+    assert(offset <= wstr.length());
+    if (offset < wstr.length()) {
+      row_tokens.emplace_back(ws, wstr.length() - offset);
```

Comment:
Prefer push_back (reference https://abseil.io/tips/112)

Reply:
The article gives an example of a std::string as where emplace would be a fare game. :)

---
In reply to: [238896646](https://github.com/Microsoft/onnxruntime/pull/31#discussion_r238896646) [](ancestors = 238896646)

Reply:
I guess in this case it's probably ok. You *want* a vector to be pushed back.